src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonDictionaryConverter.cs

```diff
@@ -270,27 +277,22 @@ protected static JsonConverter<T> GetConverter<T>(JsonTypeInfo typeInfo)
             value = (TDictionary)state.Current.ReturnValue!;
             return true;
 
-            TKey ReadDictionaryKey(ref Utf8JsonReader reader, ref ReadStack state)
+            static TKey ReadDictionaryKey(JsonConverter<TKey> keyConverter, ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
             {
                 TKey key;
-                string unescapedPropertyNameAsString;
-
-                _keyConverter ??= GetConverter<TKey>(state.Current.JsonTypeInfo.KeyTypeInfo!);
+                string unescapedPropertyNameAsString = reader.GetString()!;
+                state.Current.JsonPropertyNameAsString = unescapedPropertyNameAsString; // Copy key name for JSON Path support in case of error.
```

Comment:
Moves this assignment assignment _before_ the call to `ReadAsPropertyNameCore` so that the correct JsonPath is surfaced in case of an error.